### PR TITLE
Fixes #16941 - Check perms on Pulp key

### PR DESCRIPTION
**Pull request by chris1984**

### CODE REVIEW

File: manifests/config.pp [hunk 1]
```diff
@@ -86,6 +86,13 @@
     mode    => '0644',
   }
 
+  file { '/etc/pki/pulp/rsa.key':
+    ensure => file,
+    owner  => 'root',
+    group  => 'apache',
+    mode   => '0640',
+        }
```
Comment (ekohl) [inline manifests/config.pp]: Alignment is off.

**Comment by chris1984** [comment]
@ekohl it looks to be an issue with an empty file, according to this:

https://github.com/pulp/pulp/blob/master/pulp.spec#L906-L919

What are your suggestions?

**Comment by stbenjam** [comment]
This looks like where it's actually created:
https://github.com/pulp/pulp/blob/master/pulp.spec#L476-L493

I think that because pulp::config should run after pulp::install, the file will always exist by the time puppet runs here, so puppet will only fix the permissions (what we want).

**Comment by stbenjam** [comment]
To be honest though, I'm a little skeptical of this change, there are tens of thousands of files managed by packages that Foreman, Katello, and others use - puppet can't maintain permissions on them all.  

This was a one-off case of a user flailing their arms trying to fix something and making the situation worse, should we handle all of those cases? I want to say no, but also these keys do seem to be something different since they're generated in %post scriplets in the RPM and `rpm -V` doesn't seem to detect changes.  I think maybe something to do with the `%ghost` directive:

```
[root@sat-rhel7 ~]# chown root:root /etc/pki/pulp/rsa.key 
[root@sat-rhel7 ~]# rpm -V pulp-server | grep rsa.key
[root@sat-rhel7 ~]# 
```

**Comment by ekohl** [comment]
You can replicate the post in puppet:

```puppet
    exec { 'create pulp private key':
      command => "openssl genrsa -out ${::pulp::rsa_key} 2048",
      creates => $::pulp::rsa_key,
      umask   => '770',
    } ->
    exec { 'create pulp public key':
      command => "openssl rsa -in ${::pulp::rsa_key} -pubout > ${::pulp::rsa_pub}",
      creates => $::pulp::rsa_pub,
      umask   => '770',
    }
  
    file { $::pulp::rsa_key:
      ensure   => file,
      owner    => 'root',
      group    => 'apache',
      mode     => '0640',
      requires => Exec['create pulp private key'],
    }
```

**Comment by stbenjam** [comment]
I wouldn't replicate the the %post scriplet, what if it changes in the future?

**Comment by ekohl** [comment]
Maybe an even simpler solution: if you remove the ```ensure => file``` puppet will only ensure the permissions and not create it.

File: manifests/config.pp [hunk 2]
```diff
@@ -86,6 +86,13 @@
     mode    => '0644',
   }
 
+  file { '/etc/pki/pulp/rsa.key':
```
Comment (ekohl) [inline manifests/config.pp]: We already have this path as ```$::pulp::rsa_key``` so that would be preferred.

File: manifests/config.pp [hunk 3]
```diff
@@ -86,6 +86,13 @@
...
+    ensure => file,
```
Comment (ekohl) [inline manifests/config.pp]: If you remove this, then puppet won't create the file and thus not interfere with the ```%post```.

**Comment by stbenjam** [comment]
> Maybe an even simpler solution: if you remove the ensure => file puppet will only ensure the permissions and not create it.

That seems better, but I'm not sure I want to get in the business of managing permissions on packaged files.

**Comment by ekohl** [comment]
IMHO this is a bit of a grey area. It is in post so it's not really managed, just a best effort. It's also in /etc so I could accept this (given the changes I suggested) but I don't know how common this is. I'll leave it up to you @stbenjam.

**Comment by stbenjam** [comment]
It is tracked in the RPM spec files, with a permission on it: 
  https://github.com/pulp/pulp/blob/master/pulp.spec#L442 (see like 439 above)

I feel like it's more a bug in RPM that it doesn't track this in verify (-V).

@chris1984 But, it's fine - if you want to remove the `ensure => file` as @ekohl suggested we can take this.

**Comment by stbenjam** [comment]
Thanks @chris1984

**Pull request merged**